### PR TITLE
feat(secrets):  Add support for externally created secrets

### PR DESCRIPTION
**Pull request by adutchak**
Add support for externally created secrets

### CODE REVIEW

File: api/v1alpha1/avalanchego_types.go [hunk 1]
```diff
@@ -47,6 +47,10 @@ type AvalanchegoSpec struct {
 	// +optional
 	Genesis string `json:"genesis,omitempty"`
 
+	// Predefined secrets for nodes, quantity, should correlate to nodeCount
+	// +optional
+	Secrets []string `json:"secrets,omitempty"`
```
Comment (yevhenvolchenko) [inline api/v1alpha1/avalanchego_types.go:52]: ExistingSecrets?

File: controllers/avalanchego_controller.go [hunk 2]
```diff
@@ -89,18 +94,48 @@ func (r *AvalanchegoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		}
 		return ctrl.Result{}, err
 	}
+	//Number of secrets should match nodeCount
```
Comment (yevhenvolchenko) [inline controllers/avalanchego_controller.go:97]: please add `TODO: move to validation webhook`

File: controllers/avalanchego_controller.go [hunk 3]
```diff
@@ -89,18 +94,48 @@ func (r *AvalanchegoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	// Genesis should be inside secrets or omitted for mainnet, fuji or local networks
+	if len(instance.Spec.Secrets) > 0 && instance.Spec.Genesis != "" {
+		err = errors.NewBadRequest("Genesis cannot be specified when using pre-defined secrets. genesis.json key should be avaliable in secret instead and AVAGO_GENESIS env var provided.")
+		instance.Status.Error = err.Error()
+		if err := r.Status().Update(ctx, instance); err != nil {
+			l.Error(err, "error calling Update")
+		}
+		return ctrl.Result{}, err
+	}
+	// Certificates should be inside secrets
```
Comment (yevhenvolchenko) [inline controllers/avalanchego_controller.go:115]: please add `TODO: move to validation webhook`

File: controllers/avalanchego_controller.go [hunk 4]
```diff
@@ -89,18 +94,48 @@ func (r *AvalanchegoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+			l.Error(err, "error calling Update")
+		}
+		return ctrl.Result{}, err
+	}
 
 	// Ignore these environment variables if given
 	for i, v := range instance.Spec.Env {
 		switch v.Name {
-		case "AVAGO_PUBLIC_IP", "AVAGO_HTTP_HOST", "AVAGO_STAKING_TLS_CERT_FILE", "AVAGO_STAKING_TLS_KEY_FILE", "AVAGO_DB_DIR", "AVAGO_HTTP_PORT", "AVAGO_STAKING_PORT", "AVAGO_GENESIS":
+		case "AVAGO_PUBLIC_IP", "AVAGO_HTTP_HOST", "AVAGO_STAKING_TLS_CERT_FILE", "AVAGO_STAKING_TLS_KEY_FILE", "AVAGO_DB_DIR", "AVAGO_HTTP_PORT", "AVAGO_STAKING_PORT":
```
Comment (yevhenvolchenko) [inline controllers/avalanchego_controller.go:128]: "AVAGO_GENESIS" should still be discarded

File: controllers/avalanchego_controller_ensure.go [hunk 5]
```diff
@@ -70,7 +70,8 @@ func (r *AvalanchegoReconciler) ensureSecret(
 ) error {
 	isSecretUpdateable := isNotUpdateable
 
-	if instance.Spec.Genesis != "" && len(instance.Spec.Certificates) != 0 {
+	// Check len(instance.Spec.Secrets) == 0 means client already created pre-defined secrets. They should not be updated.
+	if instance.Spec.Genesis != "" && len(instance.Spec.Certificates) != 0 && len(instance.Spec.Secrets) == 0 {
```
Comment (yevhenvolchenko) [inline controllers/avalanchego_controller_ensure.go:74]: logic seems file though

**Review by yevhenvolchenko — CHANGES_REQUESTED** [review]

File: api/v1alpha1/avalanchego_types.go (hunk 1 above)
Comment (adutchak) [reply api/v1alpha1/avalanchego_types.go]: Renamed

File: controllers/avalanchego_controller.go (hunk 3 above)
Comment (adutchak) [reply controllers/avalanchego_controller.go:118]: Added

File: controllers/avalanchego_controller_ensure.go [hunk 6]
```diff
@@ -70,7 +70,8 @@ func (r *AvalanchegoReconciler) ensureSecret(
 ) error {
 	isSecretUpdateable := isNotUpdateable
 
-	if instance.Spec.Genesis != "" && len(instance.Spec.Certificates) != 0 {
+	// Check len(instance.Spec.Secrets) == 0 means client already created pre-defined secrets. They should not be updated.
```
[the remark this answers is not in the document]
Comment (adutchak) [reply controllers/avalanchego_controller_ensure.go]: Corrected

File: controllers/avalanchego_controller.go [hunk 7]
```diff
@@ -89,18 +94,48 @@ func (r *AvalanchegoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+	//Number of secrets should match nodeCount
+	if len(instance.Spec.Secrets) > 0 && len(instance.Spec.Secrets) != instance.Spec.NodeCount {
+		err = errors.NewBadRequest("Number of provided secrets does not match nodeCount")
+		instance.Status.Error = err.Error()
+		if err := r.Status().Update(ctx, instance); err != nil {
+			l.Error(err, "error calling Update")
+		}
+		return ctrl.Result{}, err
+	}
+	// Genesis should be inside secrets or omitted for mainnet, fuji or local networks
```
[the remark this answers is not in the document]
Comment (adutchak) [reply controllers/avalanchego_controller.go:108]: Added

**Review by yevhenvolchenko — APPROVED** [review]

**Pull request merged**